renku/core/models/provenance/datasets.py

```diff
@@ -377,13 +443,30 @@ def add(self, dataset):
 
     def get(self, identifier):
         """Return a dataset by its original identifier."""
-        datasets = [d for d in self._datasets if d.identifier == identifier]
-        assert len(datasets) <= 1, f"Found more than one with identifier `{identifier}`."
-        return datasets[0] if datasets else None
+        datasets = (d for d in self._datasets if d.identifier == identifier)
+        dataset = next(datasets, None)
+        assert next(datasets, None) is None, f"Found more than one dataset with identifier `{identifier}`."
+        return dataset
 
     def get_by_name(self, name):
-        """Return a list of datasets by name."""
-        return [d for d in self._datasets if d.name == name]
+        """Return a generator that yields datasets by name."""
+        return (d for d in self._datasets if d.name == name)
+
+    def get_latest_by_name(self, name):
+        """Return the latest version of a dataset."""
+        datasets_ids = set(d.id for d in self.get_by_name(name))
+
+        for dataset in self.get_by_name(name):
+            if dataset.derived_from:
+                datasets_ids.remove(dataset.derived_from.url_id)
+
+        assert len(datasets_ids) <= 1, f"There are more than one latest versions with name `{name}`"
+
+        if not datasets_ids:
+            return None
+
+        url = datasets_ids.pop()
+        return next(d for d in self.get_by_name(name) if d.id == url)
```

Comment:
```suggestion
        datasets = {d.id: d for d in self.get_by_name(name)}

        for dataset in datasets.values():
            if dataset.derived_from:
                datasets.pop(dataset.derived_from.url_id, None)

        assert len(datasets) <= 1, f"There are more than one latest versions with name `{name}`"

        if not datasets:
            return None

        _, dataset = datasets.popitem()
        return dataset
```

would reduce the number of times we need to iterate over this.

Reply:
We cannot iterate over a dict and remove items from it; I've just changed `for dataset in datasets.values()` to `for dataset in self.get_by_name(name)`

Reply:
right, `for dataset in list(datasets.values())` would work though. I'm just thinking in a project with a large provenance, we want to loop over the provenance as little as possible.